lib/spack/spack/cmd/license.py

```diff
@@ -89,50 +90,101 @@ def list_files(args):
         print(os.path.join(spack.paths.spack_root, relpath))
 
 
+class LicenseError(object):
+    OLD_LICENSE = 1
+    SPDX_MISMATCH = 2
+    GENERAL_MISMATCH = 3
+
+    def __init__(self):
+        self.error_counts = defaultdict(int)
+
+    def add_error(self, error):
+        self.error_counts[error] += 1
+
+    def has_errors(self):
+        return sum(self.error_counts.values()) > 0
+
+    def error_messages(self):
+        total = sum(self.error_counts.values())
+        missing = self.error_counts[LicenseError.GENERAL_MISMATCH]
+        spdx_mismatch = self.error_counts[LicenseError.SPDX_MISMATCH]
+        old_license = self.error_counts[LicenseError.OLD_LICENSE]
+        return (
+            '%d improperly licensed files' % (total),
+            'files with wrong SPDX-License-Identifier:   %d' % spdx_mismatch,
+            'files with old license header:              %d' % old_license,
+            'files not containing expected license:      %d' % missing)
+
+def _check_license(lines, path):
+    license_lines = [
+        r'Copyright 2013-(?:201[789]|202\d) Lawrence Livermore National Security, LLC and other',  # noqa: E501
```

Comment:
Make it ensure the year of the most recent commit, with a grace period of January (so if the commit is in January 2020, 2019 is still an ok year, so we have a month to get it updated).  Taking commit date ensures that past commits will continue to pass tests when checked out.

Reply:
So to be clear it should fail if the date is not `year of latest commit or (year-1 if latest commit is january)`?

To be clear, right now it is less strict: it only prints a debug message if the end date doesn't match. About 20 files will have to change if "2018" is an error (which is fine, I'm just clarifying).

Reply:
Someone could run this in a Spack instance stripped of its git history, I don't think we want to tie it to a commit.

Reply:
I'll note that doing any checking at all on the date is entirely new, and possibly worth handling in another PR, especially since both of you make good points: for example perhaps it should try to use git, and be more lenient if git is not available. Another possibility is that it could parse the date from a single "official" file that's part of the Spack core, so as long as we remember to update that, then the packages will be handled (and all without invoking git).

Side note: running `spack license` requires `git` to be available, but git is never actually run, so Spack instances stripped of git history should be able to run `spack license` (that's not to say it *must* be supported, but ).